lib/winapi/filemanip.c

```diff
@@ -241,6 +242,130 @@ BOOL MoveFileA (LPCTSTR lpExistingFileName, LPCTSTR lpNewFileName)
+            FILE_SHARE_READ,
+            FILE_READ_ATTRIBUTES | FILE_SYNCHRONOUS_IO_NONALERT | FILE_OPEN_FOR_BACKUP_INTENT | FILE_NO_EA_KNOWLEDGE);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        return FALSE;
+    }
+
+    // Fail overwriting the file if the target == the source. This is
+    // intentionally done after the check for existence of the source
+    // to set the most appropriate LastError value.
+    if (!strcmp(lpExistingFileName, lpNewFileName)) {
+      NtClose(sourceHandle);
+      SetLastError(ERROR_SHARING_VIOLATION);
+      return FALSE;
+    }
+
+    status = NtQueryInformationFile(
+            sourceHandle,
+            &ioStatusBlock,
+            &networkOpenInformation,
+            sizeof(networkOpenInformation),
+            FileNetworkOpenInformation);
+    if (!NT_SUCCESS(status)) {
+        DWORD lastError = RtlNtStatusToDosError(status);
+        SetLastError(lastError);
+        return FALSE;
+    }
+
+    RtlInitAnsiString(&targetPath, lpNewFileName);
+    InitializeObjectAttributes(&objectAttributes, &targetPath, OBJ_CASE_INSENSITIVE, ObDosDevicesDirectory(), NULL);
+
+    status = NtCreateFile(
+            &targetHandle,  // FileHandle
+            FILE_GENERIC_WRITE,   // DesiredAccess
+            &objectAttributes, // ObjectAttributes
+            &ioStatusBlock,   // IoStatusBlock
+            &networkOpenInformation.AllocationSize,  // AllocationSize
+            networkOpenInformation.FileAttributes,  // FileAttributes
+            FILE_SHARE_READ | FILE_SHARE_WRITE,  // ShareAccess
+            bFailIfExists ? FILE_CREATE : FILE_SUPERSEDE,  // CreateDisposition
+            FILE_SYNCHRONOUS_IO_NONALERT | FILE_OPEN_FOR_BACKUP_INTENT);  // CreateOptions
+    if (!NT_SUCCESS(status)) {
+        NtClose(sourceHandle);
+        DWORD lastError = RtlNtStatusToDosError(status);
+        SetLastError(lastError);
+        return FALSE;
+    }
+
+    readBuffer = MmAllocateContiguousMemory(readBufferSize);
+    if (!readBuffer) {
+        NtClose(sourceHandle);
+        NtClose(targetHandle);
+        SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+        return FALSE;
+    }
+
+    while (TRUE) {
+        const BYTE *bufferPos = readBuffer;
+        if (!ReadFile(sourceHandle, readBuffer, readBufferSize, &bytesRead, NULL)) {
+            MmFreeContiguousMemory(readBuffer);
+            NtClose(sourceHandle);
+            NtClose(targetHandle);
+            return FALSE;
+        }
+
+        if (!bytesRead) {
+            break;
+        }
+
+        while (bytesRead > 0) {
+            DWORD bytesWritten = 0;
+            if (!WriteFile(targetHandle, bufferPos, bytesRead, &bytesWritten, NULL)) {
+                MmFreeContiguousMemory(readBuffer);
+                NtClose(sourceHandle);
+                NtClose(targetHandle);
+                return FALSE;
+            }
+            bytesRead -= bytesWritten;
+            bufferPos += bytesWritten;
+        }
```

Comment:
Doesn't the kernel offer some call or IOCTL to copy files?
This user-space copy loop feels a bit involved and there's probably a lot that can go wrong (stuff like not applying the proper flags to the file when creating it, not taking advantage of file-system optimizations etc.)

Reply:
Totally agree; I didn't see a function that looked appropriate in `xboxkrnl.h` so I assumed there was no lower level method that could be utilized. Similarly I did not have luck searching the win32 docs.

Is there someplace else I could look?

Reply:
Poked around a bit more, the only lower level copying I found was added as part of windows server after win2k/xp days.

Reply:
Possibly something more like `NtWriteFile()` than using the user abstraction `WriteFile()` is what he's hinting at?

Reply:
Ah, I assume he was looking for something more like https://docs.microsoft.com/en-us/windows/win32/api/winioctl/ni-winioctl-fsctl_duplicate_extents_to_file where the entire copy loop is handled at a lower level (but that's not available in the XBOX kernel). 

I think a switch to `NtWriteFile` would end up duplicating much of the content of `WriteFile` and still operating in user-space. I don't know enough to say if the efficiency gain warrants the maintenance cost, but happy to make that change if it's preferred.

@JayFoxRox can you clarify?